core/src/main/java/org/jdbi/v3/core/array/SqlArrayArgumentFactory.java

```diff
@@ -47,8 +47,12 @@
                     .map(arrayType -> new SqlArrayArgument<>(arrayType, value));
         }
 
-        return GenericTypes.findGenericParameter(type, Collection.class)
-                .flatMap(lookup)
+        final Collection<?> collection = (Collection<?>) value;
+        // If there's no static type, guess based on the first element
+        final Type eltType = GenericTypes.findGenericParameter(type, Collection.class)
+            .orElse(collection.isEmpty() ? String.class : collection.iterator().next().getClass());
```

Comment:
A couple things here:

* Let's use `orElseGet` so we don't produce a guess when we can compute the element type.
* We should go by the first _non-null_ element, otherwise we risk an NPE:

```java
Type eltType = GenericTypes.findGenericParameter(type, Collection.class)
    .orElseGet(() ->
        collection.stream()
            .filter(Objects::nonNull)
            .map(Object::getClass)
            .findFirst()
            .orElse(null));
```

* I think assuming `String` as the element type when the collection is empty or contains only nulls is a bad idea. If the collection might by empty, then users should call `bindByType`, and give JDBI a generic signature to go off of. This is consistent with the requirement that they call `bindByType` for any null parameter of a custom data type. If this is a concern, I wouldn't be opposed to a set of `bindCollection` methods:

```java
public final <T> This bindCollection(int position,
                                     Collection<T> collection,
                                     Type<T> elementType) { ... }

public final <T> This bindCollection(int position,
                                     Collection<T> collection,
                                     GenericType<T> elementType) { ... }

public final <T> This bindCollection(String name,
                                     Collection<T> collection,
                                     Type<T> elementType) { ... }

public final <T> This bindCollection(String name,
                                     Collection<T> collection,
                                     GenericType<T> elementType) { ... }
```